### PR TITLE
Code quality

### PR DESCRIPTION
**Pull request by trueqbit**
* Used feature detection macros for polyfills
* Fixed spelling of "limit_accesor" -> "limit_accessor"
* Included dependent headers
* Removed some superfluous constructors that are just forwarding their arguments to initialize members
* Qualified enum values by their original enum class name instead of using `decltype`
* Spelled out type for SQLite database connection instead of using `auto` (I am not a friend of auto everywhere, because it makes code harder to understand)
* Moved `storage_t<>::drop_trigger()` to base class `storage_base`, changed signature to accept a `std::string`
* Addressed warning when streaming `conflict_action`
* Used `polyfill::negation` instead of `static_not`
* `mapped_type_proxy_t`
* Simplified member traits
* Gone is all the boiler-plate code when preparing, binding to, and executing statements

### CODE REVIEW

File: dev/storage_base.h [hunk 1]
```diff
@@ -47,19 +45,24 @@ namespace sqlite_orm {
             void drop_index(const std::string& indexName) {
                 std::stringstream ss;
                 ss << "DROP INDEX " << quote_identifier(indexName);
-                perform_void_exec(get_connection().get(), ss.str());
+                perform_void_exec(this->get_connection().get(), ss.str());
+            }
+
+            void drop_trigger(const std::string& triggerName) {
```
Comment (trueqbit) [inline dev/storage_base.h:51]: @fnc12 This was a [templated method](https://github.com/fnc12/sqlite_orm/blob/451a81b774123d07f45413db4f4604aae0773fb1/dev/storage.h#L217) previously. Was there any reason for this?

File: dev/storage_base.h [hunk 2]
```diff
@@ -328,7 +331,6 @@ namespace sqlite_orm {
                 std::stringstream ss;
                 ss << C::name();
                 auto name = ss.str();
-                ss.flush();
```
Comment (fnc12) [inline dev/storage_base.h:331]: removing this will flush in the end of the scope -> after `this->create_collation(name, move(func));`. Better to drop unused memory ASAP

File: dev/storage_base.h [hunk 3]
```diff
@@ -360,7 +362,6 @@ namespace sqlite_orm {
...
```
Comment (fnc12) [inline dev/storage_base.h:363]: same here

File: dev/table_info.h [hunk 4]
```diff
@@ -12,6 +14,7 @@ namespace sqlite_orm {
         std::string dflt_value;
         int pk = 0;
 
+#ifndef SQLITE_ORM_AGGREGATE_PAREN_INIT
```
Comment (fnc12) [inline dev/table_info.h:17]: what is the purpose of `SQLITE_ORM_AGGREGATE_PAREN_INIT`? It covers all trivial constructors?

File: patch1.diff [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+---
+ dev/statement_serializer.h | 2 ++
+ 1 file changed, 2 insertions(+)
+
+diff --git a/dev/statement_serializer.h b/dev/statement_serializer.h
+index 10c2a67..f2da5e5 100644
+--- a/dev/statement_serializer.h
++++ b/dev/statement_serializer.h
+@@ -1774,6 +1774,8 @@ namespace sqlite_orm {
```
Comment (fnc12) [inline patch1.diff:14]: looks like a redundant file

**Comment by fnc12** [comment]
@trueqbit please remove `.diff` files

**Review by fnc12 — CHANGES_REQUESTED** [review]

File: dev/storage_base.h (hunk 1 above)
Comment (trueqbit) [reply dev/storage_base.h:51]: I am just wondering then why the other methods don't accept "any thing"?

File: dev/cxx_polyfill.h [hunk 6]
```diff
@@ -27,50 +45,31 @@ namespace sqlite_orm {
...
-
-            template<template<typename...> class Primary, class... Types>
-            SQLITE_ORM_INLINE_VAR constexpr bool is_specialization_of_v<Primary<Types...>, Primary> = true;
-
-            template<typename Type, template<typename...> class Primary>
-            struct is_specialization_of : bool_constant<is_specialization_of_v<Type, Primary>> {};
-
-            template<typename... T>
-            using is_specialization_of_t = typename is_specialization_of<T...>::type;
+#if 0  // __cpp_lib_detect >= 0L  //  library fundamentals TS v2, [meta.detect]
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/cxx_polyfill.h:66]: Because there's no known date yet, it's still experimental. There's `__cpp_lib_experimental_detect`, however I don't want to use experimental because usually it's going to be removed at some point in the future.

File: dev/cxx_polyfill.h [hunk 7]
```diff
@@ -108,6 +107,24 @@ namespace sqlite_orm {
             SQLITE_ORM_INLINE_VAR constexpr bool is_detected_v = is_detected<Op, Args...>::value;
 #endif
 
+#if 1  // proposed but not pursued                                                                                     \
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/cxx_polyfill.h:110]: There's no feature detection macro, however I wanted to express that it's something to be filled in later.

File: dev/statement_serializer.h [hunk 8]
```diff
@@ -1774,6 +1774,8 @@ namespace sqlite_orm {
                         return "IGNORE";
                     case conflict_action::rollback:
                         return "ROLLBACK";
+                    default:
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/statement_serializer.h:1777]: Why exactly? Not having it also produces a warning with msvc.

File: dev/table_info.h (hunk 4 above)
Comment (trueqbit) [reply dev/table_info.h:17]: I am not sure I exactly understand what you mean by whether it covers all trivial constructors.
C++20 allows to initialize aggregates via parentheses, so e.g. `table_info()` works as well as `vector<table_info>::.emplace_back()`. So this constructor is no longer needed.

File: patch1.diff (hunk 5 above)
Comment (trueqbit) [reply patch1.diff:14]: gitkraken has the annoying feature of automatically staging new files.

File: dev/conditions.h [hunk 9]
```diff
@@ -421,11 +421,6 @@ namespace sqlite_orm {
         struct order_by_base {
             int asc_desc = 0;  //  1: asc, -1: desc
             std::string _collate_argument;
-
-            order_by_base() = default;
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/conditions.h:425]: I suggest not to keep legacy around? If it's a problem in a certain version of Visual C++ I rather surround it using `#if defined(_MSC_VER) && (_MSC_VER < 1920)` or alike. Only very old VC++ versions had some problems like with `is_base_of_template`.

File: dev/storage_base.h (hunk 1 above)
Comment (fnc12) [reply dev/storage_base.h:51]: cause it is legacy.
yes `std::string_view` is a perfect solution but it is not available in C++14.

File: dev/cxx_polyfill.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (fnc12) [reply dev/cxx_polyfill.h:66]: so is it ok that we'll commit `#if 0` into `dev`?

File: dev/cxx_polyfill.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (fnc12) [reply dev/cxx_polyfill.h:110]: are you going to merge in as is into `dev` or is it still in progress?

File: dev/statement_serializer.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (fnc12) [reply dev/statement_serializer.h:1781]: hm alright. Let it be that way cause VS is more popular then Xcode and I reckon I can disable this warning in Xcode

File: dev/storage_base.h (hunk 2 above)
Comment (fnc12) [reply dev/storage_base.h:331]: not new, it is very old but yes it is not done consistently cause I added it only when I remembered about it.
> It's more code for little gain. 
this is how C and C++ work 🙂. It is not so critical cause no one asked for it yet. So it is up to you

File: dev/table_info.h (hunk 4 above)
Comment (fnc12) [reply dev/table_info.h]: I mean do we need to use `SQLITE_ORM_AGGREGATE_PAREN_INIT` for defining ctors in every struct?

File: patch1.diff (hunk 5 above)
Comment (fnc12) [reply patch1.diff]: try GitHub desktop (or git bash)

File: dev/conditions.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (fnc12) [reply dev/conditions.h:425]: hm alright. Let's wait until one more developer comes with such an issue and try to fix it differently. Really I created and deleted such ctors several times before again in again.

File: dev/table_info.h (hunk 4 above)
Comment (trueqbit) [reply dev/table_info.h]: No, only those that will be constructed using `make_unique` or `emplace_back`, which call the parentheses constructor instead of the brace initializer. In this specific case, table_info gets emplaced into the container. Which in turn is the only one of that kind in sqlite_orm.

File: dev/storage_base.h (hunk 2 above)
Comment (trueqbit) [reply dev/storage_base.h:331]: No worries, I made it state-of-the-art with a later commit in this PR.

File: dev/statement_serializer.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/statement_serializer.h:1781]: Where do you get warnings? I'll try with clang. Let's not replace one warning with another.

File: dev/cxx_polyfill.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/cxx_polyfill.h:110]: It's going into dev like this - note that nothing changed.
However if you have a better suggestion, feel free to let me know :)

File: dev/cxx_polyfill.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/cxx_polyfill.h:66]: It's going into dev like this - note that nothing changed (it's just the inverse of the previous `#if 0`.
However if you have a better suggestion, feel free to let me know :)

File: dev/conditions.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/conditions.h:425]: Yep, I have seen 3 tickets in regard to VC 2015 Update 3. For now I have reinstated those constructors using a proper feature detection macro (support for "Aggregate classes with default member initializers").

File: dev/storage_base.h (hunk 1 above)
Comment (trueqbit) [reply dev/storage_base.h:51]: I assume that `drop_trigger` came after `drop_index` and `drop_table`? If so, can we just adjust `drop_trigger` to the other methods, and possibly make a feature request?

File: dev/statement_serializer.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (fnc12) [reply dev/statement_serializer.h:1781]: on the `default` line saying 'this code will never be executed'

**Comment by fnc12** [comment]
IDK why I cannot comment right inside conversation so I'll answer there with a quote.

> I assume that drop_trigger came after drop_index and drop_table? If so, can we just adjust drop_trigger to the other methods, and possibly make a feature request?

What do you mean by feature request? Feature request with what?

File: dev/storage_base.h (hunk 1 above)
Comment (trueqbit) [reply dev/storage_base.h:51]: If somebody wants to pass a different thing than a std::string, that would be a feature request. Either we create that request ourselves or somebody in need.

Comment (fnc12) [reply dev/storage_base.h:51]: oh I see. Ok let's leave it like that. Probably we are discussing about something not used at all.

File: dev/implementations/storage_impl_definitions.h [hunk 10]
```diff
@@ -21,11 +21,10 @@ namespace sqlite_orm {
             bool result = false;
             std::stringstream ss;
             ss << "SELECT COUNT(*) FROM sqlite_master WHERE type = " << quote_string_literal("table"s)
-               << " AND name = " << quote_string_literal(tableName);
-            auto query = ss.str();
+               << " AND name = " << quote_string_literal(tableName) << std::flush;
             auto rc = sqlite3_exec(
                 db,
-                query.c_str(),
+                ss.str().c_str(),
```
Comment (fnc12) [inline dev/implementations/storage_impl_definitions.h:27]: tip: stopping keeping `query` as a dedicated variable makes it harder to debug

File: tests/static_tests/node_tuple.cpp [hunk 11]
```diff
@@ -768,7 +768,7 @@ TEST_CASE("Node tuple") {
         }
 #endif
         SECTION("coalesce") {
-            auto f = char_(10, 20);
+            auto f = coalesce(10, 20);
```
Comment (fnc12) [inline tests/static_tests/node_tuple.cpp:771]: lol thanks

File: tests/storage_tests.cpp [hunk 12]
```diff
@@ -259,7 +269,7 @@ TEST_CASE("issue880") {
                                 foreign_key(&X::fkey_Rendimiento).references(&Rendimiento::id)));
     storage.sync_schema();
 
-    Fondo fondo{5};
```
Comment (fnc12) [inline tests/storage_tests.cpp:262]: why did you remove 5?

**Review by fnc12 — APPROVED** [review]
wow this is a lot of good job done here! I am impressed!

File: tests/storage_tests.cpp (hunk 12 above)
Comment (trueqbit) [reply tests/storage_tests.cpp:262]: What's the point?

File: tests/CMakeLists.txt [hunk 13]
```diff
@@ -133,6 +133,11 @@ endif()
 
 if (MSVC)
 	add_compile_options(/MP) # Enable multi parrallel build
+	if (MSVC_VERSION LESS_EQUAL 1900)
```
Comment (trueqbit) [inline tests/CMakeLists.txt:136]: I don't see the reason why appveyor doesn't pick up this condition and does not disable those specific warnings? [See appveyor build](https://ci.appveyor.com/project/fnc12/sqlite-orm/builds/43393529/job/r9kh99ga09hrqhm8).

File: tests/storage_tests.cpp (hunk 12 above)
Comment (fnc12) [reply tests/storage_tests.cpp:262]: 5 is checked at runtime. If you remove 5 it will be equal to 0. I know that the result is not checked and it will not alter UT results. But I don't see the reason to change this. Anyway it is just a question

File: appveyor.yml [hunk 14]
```diff
@@ -27,13 +24,19 @@ platform:
   
 environment:
   matrix:
+   - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2015
+     SQLITE_ORM_CXX_STANDARD: ""
+
    # using c++14
-   - SQLITE_ORM_CXX_STANDARD: "-DSQLITE_ORM_ENABLE_CXX_14=ON"
+   - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2022
```
Comment (trueqbit) [inline appveyor.yml:31]: There's still something wrong with the environment matrix: appveyor shouldn't build using Visual Studio 2022 with no `SQLITE_ORM_CXX_STANDARD` at all. [See appvyeor build](https://ci.appveyor.com/project/fnc12/sqlite-orm/builds/43393952).

Comment (fnc12) [reply appveyor.yml:31]: stranger things. Let's try to figure it out

File: tests/CMakeLists.txt (hunk 13 above)
Comment (fnc12) [reply tests/CMakeLists.txt:136]: Actually I am not a master in CMake and CI configs but let me try to read it and understand how it works

**Comment by fnc12** [comment]
everything works on my mac but CI got to be fixed

File: tests/storage_tests.cpp (hunk 12 above)
Comment (trueqbit) [reply tests/storage_tests.cpp:262]: The reason was simply not having to provide a conditional constructor (due to a compiler's possibly missing default member initializer support).
But you have a valid point. Let me change it back.

File: tests/static_tests/node_tuple.cpp (hunk 11 above)
Comment (trueqbit) [reply tests/static_tests/node_tuple.cpp:771]: one unit test more, yay!

File: dev/implementations/storage_impl_definitions.h (hunk 10 above)
Comment (trueqbit) [reply dev/implementations/storage_impl_definitions.h:27]: I thought that the code is simple enough. Note that the stream gets flushed, so you will see the full resulting string in the stringstream variable.

File: appveyor.yml (hunk 14 above)
Comment (trueqbit) [reply appveyor.yml:31]: I am looking at appveyor's [build environment doco](https://www.appveyor.com/docs/build-environment/#using-multiple-images-for-the-same-build), which I would think is straightforward.

Comment (trueqbit) [reply appveyor.yml:31]: It's obvious once you see things through the eyes of a yaml parser: `- SQLITE_ORM_CXX_STANDARD` shouldn't be on the same level with a dash :roll_eyes:

File: dev/object_from_column_builder.h [hunk 15]
```diff
@@ -27,10 +32,10 @@ namespace sqlite_orm {
                 object_from_column_builder_base{stmt_}, object(object_) {}
 
             template<class C>
-            void operator()(const C& c) const {
+            void operator()(const C& c) {
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/object_from_column_builder.h:35]: Yes, because a 'lambda' passed to the tuple iteration function doesn't need to be const anymore.

File: dev/backup.h [hunk 16]
```diff
@@ -32,14 +33,12 @@ namespace sqlite_orm {
...
-                handle(other.handle), holder(move(other.holder)), to(other.to), from(other.from) {
-                other.handle = nullptr;
-            }
+                handle(std::exchange(other.handle, nullptr)), holder(move(other.holder)), to(other.to),
+                from(other.from) {}
 
             ~backup_t() {
                 if(this->handle) {
                     (void)sqlite3_backup_finish(this->handle);
-                    this->handle = nullptr;
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/backup.h:42]: I've never seen good code that relies on 'nulling' out members in a destructor, and I am not aware of any standard library implementer doing it. I have never needed this kind of debugging technique ever in my programmer life. If you want we can make it conditional via `#ifndef NDEBUG`?

File: dev/ast_iterator.h [hunk 17]
```diff
@@ -39,18 +39,18 @@ namespace sqlite_orm {
              *  L is a callable type. Mostly is a templated lambda
              */
             template<class L>
-            void operator()(const T& t, const L& l) const {
-                l(t);
+            void operator()(const T& t, L& lambda) const {
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/ast_iterator.h:42]: Because here it just preserves the const/non-const qualification.

File: dev/static_magic.h [hunk 18]
```diff
@@ -17,27 +17,24 @@ namespace sqlite_orm {
         }
 
         template<typename T, typename F>
-        decltype(auto) static_if(std::true_type, const T& t, const F&) {
-            return (t);
+        decltype(auto) static_if(std::true_type, T&& t, const F&) {
+            return static_cast<T&&>(t);
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/static_magic.h:21]: As written above, just for preserving the way it was passed in.

File: dev/table.h [hunk 19]
```diff
@@ -116,10 +119,10 @@ namespace sqlite_orm {
              *  Calls **l** with every primary key dedicated constraint
              */
             template<class L>
-            void for_each_primary_key(const L& lambda) const {
+            void for_each_primary_key(L&& lambda) const {
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/table.h:122]: Here we can do that, I updated it in line with the other 'lambda' changes, such that we don't care here what the function object does.

File: dev/backup.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (fnc12) [reply dev/backup.h:42]: ok let's leave it like this

File: dev/ast_iterator.h [hunk 20]
```diff
@@ -683,8 +683,8 @@ namespace sqlite_orm {
             using node_type = collate_t<T>;
 
             template<class L>
-            void operator()(const node_type& node, const L& l) const {
-                iterate_ast(node.expr, l);
+            void operator()(const node_type& node, L& lambda) const {
```
[the remark this answers is not in the document]
Comment (fnc12) [reply dev/ast_iterator.h:686]: alright. I see the meaning. Let's leave it like this. Thanks

**Comment by fnc12** [comment]
Снимок экрана 2022-05-05 в 11 29 55

**Review by fnc12 — APPROVED** [review]

**Comment by trueqbit** [comment]
> C++14
Now I am surprised! Even VC 2015 already shipped `std::negate`.

**Comment by fnc12** [comment]
Снимок экрана 2022-05-06 в 19 32 09
how it worked on CI?

**Comment by trueqbit** [comment]
> Снимок экрана 2022-05-06 в 19 32 09
> how it worked on CI?

Because msvc doesn't yet report shadowing template paramters (I may have missed a compiler option).

**Comment by fnc12** [comment]
thanks @trueqbit
